fml/synchronization/sync_switch.cc

```diff
@@ -0,0 +1,43 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/fml/synchronization/sync_switch.h"
+
+namespace fml {
+
+SyncSwitch::Handlers::Handlers() : true_handler({}), false_handler({}) {}
+
+SyncSwitch::Handlers& SyncSwitch::Handlers::SetTrue(
+    const std::function<void()>& handler) {
+  true_handler = handler;
+  return *this;
+}
+
+SyncSwitch::Handlers& SyncSwitch::Handlers::SetFalse(
+    const std::function<void()>& handler) {
+  false_handler = handler;
+  return *this;
+}
+
+SyncSwitch::SyncSwitch() : SyncSwitch(false) {}
+
+SyncSwitch::SyncSwitch(bool value) : value_(value) {}
+
+void SyncSwitch::Execute(const SyncSwitch::Handlers& handlers) {
+  std::lock_guard<std::mutex> guard(mutex_);
```

Comment:
There is no reason to have a read-write lock here.  It would just add overhead we don't need.  I don't think it's any less correct as written, just a design choice.  I've documented the behavior in the class docstring.

Reply:
Whats the overhead? It is just a different type of mutex.

Reply:
Check out the implementations.  There is no read-write mutex primitive provided by the operating system.  Logic has to built up around the standard mutex to provide the read-write behavior.

Reply:
[`//fml/synchronization/shared_mutex.h`](https://github.com/flutter/engine/blob/master/fml/synchronization/shared_mutex.h)

Reply:
We only have that in FML because we can't use [std::shared_mutex](https://en.cppreference.com/w/cpp/thread/shared_mutex) on old versions of iOS.